File: sdk/samples/03_types/c/enums.c
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/**
 * Enums Sample - Demonstrates DDS enumeration types
 *
 * This sample shows how to work with the Enums struct:
 * - Color enum (Red=0, Green=1, Blue=2)
 * - Status enum (Unknown=0, Active=10, Inactive=20, Error=100)
 */

#include <stdio.h>
#include <string.h>
#include "generated/Enums.h"

/* Helper: Color to string */
static const char* color_to_string(Color c) {
    switch (c) {
        case COLOR_RED:   return "Red";
        case COLOR_GREEN: return "Green";
        case COLOR_BLUE:  return "Blue";
        default:          return "Unknown";
    }
}

/* Helper: Status to string */
static const char* status_to_string(Status s) {
    switch (s) {
        case STATUS_UNKNOWN:  return "Unknown";
        case STATUS_ACTIVE:   return "Active";
        case STATUS_INACTIVE: return "Inactive";
        case STATUS_ERROR:    return "Error";
        default:              return "?";
    }
}

int main(void) {
    printf("=== HDDS Enum Types Sample ===\n\n");

    uint8_t buffer[64];

    /* Color enum */
    printf("--- Color Enum ---\n");
    printf("Color values:\n");
    printf("  Red   = %d\n", COLOR_RED);
    printf("  Green = %d\n", COLOR_GREEN);
    printf("  Blue  = %d\n", COLOR_BLUE);

    /* Status enum with explicit values */
    printf("\n--- Status Enum (explicit values) ---\n");
    printf("Status values:\n");
    printf("  Unknown  = %d\n", STATUS_UNKNOWN);
    printf("  Active   = %d\n", STATUS_ACTIVE);
    printf("  Inactive = %d\n", STATUS_INACTIVE);
    printf("  Error    = %d\n", STATUS_ERROR);

    /* Enums struct with both enums */
    printf("\n--- Enums Serialization ---\n");
    Enums demo = {
        .color = COLOR_GREEN,
        .status = STATUS_ACTIVE
    };

    printf("Original:\n");
    printf("  color:  %s (%d)\n", color_to_string(demo.color), demo.color);
    printf("  status: %s (%d)\n", status_to_string(demo.status), demo.status);

    int size = enums_encode_cdr2_le(&demo, buffer, sizeof(buffer));
    if (size < 0) {
        printf("[ERROR] Serialization failed! (%d)\n", size);
        return 1;
    }
    printf("Serialized size: %d bytes\n", size);
    printf("Serialized bytes: ");
    for (int i = 0; i < size; ++i) {
        printf("%02X", buffer[i]);
    }
    printf("\n");

    Enums deser;
    memset(&deser, 0, sizeof(deser));
    enums_decode_cdr2_le(&deser, buffer, (size_t)size);
    printf("Deserialized:\n");
    printf("  color:  %s\n", color_to_string(deser.color));
    printf("  status: %s\n", status_to_string(deser.status));

    if (demo.color == deser.color && demo.status == deser.status) {
        printf("[OK] Enums round-trip successful\n\n");
    }

    /* Test all color values */
    printf("--- All Color Values Test ---\n");
    Color colors[] = {COLOR_RED, COLOR_GREEN, COLOR_BLUE};
    for (int i = 0; i < 3; ++i) {
        Enums test = {.color = colors[i], .status = STATUS_UNKNOWN};
        int test_size = enums_encode_cdr2_le(&test, buffer, sizeof(buffer));
        Enums test_deser;
        memset(&test_deser, 0, sizeof(test_deser));
        enums_decode_cdr2_le(&test_deser, buffer, (size_t)test_size);
        printf("  %s: %d -> %s\n",
               color_to_string(colors[i]), colors[i],
               color_to_string(test_deser.color));
    }
    printf("[OK] All colors round-trip correctly\n\n");

    /* Test all status values */
    printf("--- All Status Values Test ---\n");
    Status statuses[] = {STATUS_UNKNOWN, STATUS_ACTIVE, STATUS_INACTIVE, STATUS_ERROR};
    for (int i = 0; i < 4; ++i) {
        Enums test = {.color = COLOR_RED, .status = statuses[i]};
        int test_size = enums_encode_cdr2_le(&test, buffer, sizeof(buffer));
        Enums test_deser;
        memset(&test_deser, 0, sizeof(test_deser));
        enums_decode_cdr2_le(&test_deser, buffer, (size_t)test_size);
        printf("  %s: %d -> %s\n",
               status_to_string(statuses[i]), statuses[i],
               status_to_string(test_deser.status));
    }
    printf("[OK] All statuses round-trip correctly\n\n");

    /* Default values */
    printf("--- Default Values ---\n");
    Enums default_demo = {0};
    printf("Default color:  %s\n", color_to_string(default_demo.color));
    printf("Default status: %s\n", status_to_string(default_demo.status));

    printf("\n=== Sample Complete ===\n");
    return 0;
}

File: sdk/samples/03_types/python/enums.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
Enums Sample - Demonstrates DDS enumeration types

This sample shows how to work with enum types:
- Simple enums (Color)
- Enums with explicit values (Status)
"""

import sys
sys.path.insert(0, '.')

from generated.Enums import Color, Status, Enums


def main():
    print("=== HDDS Enum Types Sample ===\n")

    # Color enum
    print("--- Color Enum ---")
    print("Color values:")
    print(f"  RED   = {Color.RED.value}")
    print(f"  GREEN = {Color.GREEN.value}")
    print(f"  BLUE  = {Color.BLUE.value}")

    # Status enum with explicit values
    print("\n--- Status Enum (explicit values) ---")
    print("Status values:")
    print(f"  UNKNOWN  = {Status.UNKNOWN.value}")
    print(f"  ACTIVE   = {Status.ACTIVE.value}")
    print(f"  INACTIVE = {Status.INACTIVE.value}")
    print(f"  ERROR    = {Status.ERROR.value}")

    # Enums with both enums
    print("\n--- Enums Serialization ---")
    demo = Enums(color=Color.GREEN, status=Status.ACTIVE)

    print("Original:")
    print(f"  color:  {demo.color.name} ({demo.color.value})")
    print(f"  status: {demo.status.name} ({demo.status.value})")

    data = demo.encode_cdr2_le()
    print(f"Serialized size: {len(data)} bytes")
    print(f"Serialized bytes: {data.hex().upper()}")

    deser, _ = Enums.decode_cdr2_le(data)
    print("Deserialized:")
    print(f"  color:  {deser.color.name}")
    print(f"  status: {deser.status.name}")

    if demo == deser:
        print("[OK] Enums round-trip successful\n")

    # Test all color values
    print("--- All Color Values Test ---")
    for color in [Color.RED, Color.GREEN, Color.BLUE]:
        test = Enums(color=color, status=Status.UNKNOWN)
        test_data = test.encode_cdr2_le()
        test_deser, _ = Enums.decode_cdr2_le(test_data)
        print(f"  {color.name}: {color.value} -> {test_deser.color.name}")
        assert test == test_deser
    print("[OK] All colors round-trip correctly\n")

    # Test all status values
    print("--- All Status Values Test ---")
    for status in [Status.UNKNOWN, Status.ACTIVE, Status.INACTIVE,
                   Status.ERROR]:
        test = Enums(color=Color.RED, status=status)
        test_data = test.encode_cdr2_le()
        test_deser, _ = Enums.decode_cdr2_le(test_data)
        print(f"  {status.name}: {status.value} -> {test_deser.status.name}")
        assert test == test_deser
    print("[OK] All statuses round-trip correctly\n")

    # Default values
    print("--- Default Values ---")
    default_demo = Enums(color=Color.RED, status=Status.UNKNOWN)
    print(f"Default color:  {default_demo.color.name}")
    print(f"Default status: {default_demo.status.name}")

    print("\n=== Sample Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())
